src/test/regress/expected/recurring_outer_join.out

```diff
@@ -0,0 +1,568 @@
+DEBUG:  Wrapping relation "local_1" to a subquery
+DEBUG:  generating subplan XXX_1 for subquery SELECT a FROM recurring_outer_join.local_1 WHERE true
+DEBUG:  recursively planning left side of the right join since the outer side is a recurring rel
+DEBUG:  recursively planning distributed relation "dist_1" since it is part of a distributed join node that is outer joined with a recurring rel
+DEBUG:  Wrapping relation "dist_1" to a subquery
+DEBUG:  generating subplan XXX_2 for subquery SELECT a FROM recurring_outer_join.dist_1 WHERE true
+DEBUG:  recursively planning left side of the full join since the other side is a recurring rel
+DEBUG:  recursively planning distributed relation "dist_1" "t1" since it is part of a distributed join node that is outer joined with a recurring rel
+DEBUG:  Wrapping relation "dist_1" "t1" to a subquery
+DEBUG:  generating subplan XXX_3 for subquery SELECT a FROM recurring_outer_join.dist_1 t1 WHERE true
+DEBUG:  Plan XXX query after replacing subqueries and CTEs: SELECT count(*) AS count FROM ((SELECT t1_1.a, NULL::integer AS b FROM (SELECT intermediate_result.a FROM read_intermediate_result('XXX_3'::text, 'binary'::citus_copy_format) intermediate_result(a integer)) t1_1) t1 FULL JOIN ((SELECT dist_1_1.a, NULL::integer AS b FROM (SELECT intermediate_result.a FROM read_intermediate_result('XXX_2'::text, 'binary'::citus_copy_format) intermediate_result(a integer)) dist_1_1) dist_1 RIGHT JOIN (SELECT local_1_1.a, NULL::integer AS b FROM (SELECT intermediate_result.a FROM read_intermediate_result('XXX_1'::text, 'binary'::citus_copy_format) intermediate_result(a integer)) local_1_1) local_1 USING (a)) t2(a, b, b_1) USING (a))
+ count
+---------------------------------------------------------------------
+    43
+(1 row)
+
+  -- subqury without FROM
+  SELECT COUNT(*) FROM dist_1 t1 RIGHT JOIN (SELECT generate_series(1,10) AS a) t2 USING (a);
+DEBUG:  recursively planning left side of the right join since the outer side is a recurring rel
+DEBUG:  recursively planning distributed relation "dist_1" "t1" since it is part of a distributed join node that is outer joined with a recurring rel
+DEBUG:  Wrapping relation "dist_1" "t1" to a subquery
+DEBUG:  generating subplan XXX_1 for subquery SELECT a FROM recurring_outer_join.dist_1 t1 WHERE true
+DEBUG:  Plan XXX query after replacing subqueries and CTEs: SELECT count(*) AS count FROM ((SELECT t1_1.a, NULL::integer AS b FROM (SELECT intermediate_result.a FROM read_intermediate_result('XXX_1'::text, 'binary'::citus_copy_format) intermediate_result(a integer)) t1_1) t1 RIGHT JOIN (SELECT generate_series(1, 10) AS a) t2 USING (a))
+ count
+---------------------------------------------------------------------
+    21
+(1 row)
+
+-- queries that we still can't support --because distributed - inner side of the
+-- join references a column of the recurring - outer rel
+  -- reference table
+  SELECT COUNT(*) FROM
```

Comment:
interesting, which part is outer join here?

Reply:
Hmm, if we do the JOIN on the t2.a, the query works:

```SQL
SELECT COUNT(*)
FROM ref_1 t1
JOIN dist_1 t2 ON (t1.a = t2.a)
WHERE **t2.a** in
    (SELECT a
     FROM dist_1 t3);
```
Because, then the JOIN_SEMI is between two distributed tables instead of ref AND distributed.

I think this is a deficieny of DeferredErrorIfUnsupportedRecurringTuplesJoin. We can have similar problems even with regular left joins:
```SQL
-- same query, only the JOIN column is changed
SELECT count(*) FROM (ref_1 r1 JOIN dist_1 d1 ON (r1.a = d1.a)) LEFT JOIN dist_1 d2 ON (d2.a = r1.a);
ERROR:  complex joins are only supported when all distributed tables are co-located and joined on their distribution columns
Time: 2.591 ms
SELECT count(*) FROM (ref_1 r1 JOIN dist_1 d1 ON (r1.a = d1.a)) LEFT JOIN dist_1 d2 ON (d2.a = d1.a);
┌───────┐
│ count │
├───────┤
│    66 │
└───────┘
(1 row)
```

Reply:
iirc from yesterday's discussion, atm we're okay with that right ?